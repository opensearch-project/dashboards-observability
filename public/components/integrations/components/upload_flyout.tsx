/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React, { useState } from 'react';
import {
  EuiButton,
  EuiFilePicker,
  EuiFlexGroup,
  EuiFlexItem,
  EuiFlyout,
  EuiFlyoutBody,
  EuiFlyoutFooter,
  EuiFlyoutHeader,
  EuiForm,
  EuiFormRow,
} from '@elastic/eui';
import { useEffect } from 'react';
import { coreRefs } from '../../../../public/framework/core_refs';
import { useToast } from '../../../../public/components/common/toast';

const uploadBundle = async (bundle: File | null): Promise<Error | null> => {
  if (!bundle) {
    return new Error('No bundle selected');
  }

  const formData = new FormData();
  formData.append('file', bundle);
  try {
    const response = await coreRefs.http?.post('/api/saved_objects/_import', {
      headers: {
        // Important to leave undefined, it forces proper headers to be set for FormData
        'Content-Type': undefined,
      },
      body: formData,
      query: {
        overwrite: true,
        // If we don't disable this, we'll get 413 Content Too Large errors for many bundles
        dataSourceEnabled: false,
      },
    });

    console.log(response);
    if (response.success) {
      return null;
    } else {
      return new Error(response.body.message);
    }
  } catch (err) {
    return err;
  }
};

const checkBundle = async (bundle: File | null): Promise<boolean> => {
  if (bundle == null) {
    return false;
  }
  if (!/.*\.ndjson&/i.test(bundle.name) && bundle.type !== 'application/x-ndjson') {
    return false;
  }

  try {
    const contents = await bundle.text();
    const objects = contents
      .trim()
      .split('\n')
      .map((s) => JSON.parse(s));
    return objects.every((obj) => !obj.type || obj.type === 'integration-template');
  } catch (err) {
    return false;
  }
};

export const IntegrationUploadPicker = ({
  onFileSelected,
  isInvalid,
  setIsInvalid,
}: {
  onFileSelected: (file: File | null) => void;
  isInvalid: boolean;
  setIsInvalid: (value: boolean) => void;
}) => {
  const [blurred, setBlurred] = useState(false);
  const [checkCompleted, setCheckCompleted] = useState(false);
  const [bundle, setBundle] = useState(null as File | null);

  useEffect(() => {
    const refreshValidity = async () => {
      if (!blurred) {
        return;
      }
      const valid = await checkBundle(bundle);
      setIsInvalid(!valid);
      setCheckCompleted(true);
    };
    setCheckCompleted(false);
    refreshValidity();
    // It's important that we update *only* on the bundle changing here, so we don't report errors
    // early when the file picker is still open and get error flickering on the UI in the happy path.
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [bundle]);

  const handleFileChange = (files: FileList | null) => {
    if (files && files.length > 0) {
      setBundle(files[0]);
      onFileSelected(files[0]);
    } else {
      setBundle(null);
      onFileSelected(null);
    }
  };

  return (
    <EuiForm>
      <EuiFormRow
        label="Select file"
        isInvalid={checkCompleted && isInvalid}
        error={['Must be an ndjson bundle of integration templates']}
      >
        <EuiFilePicker
          id="integrationBundlePicker"
          initialPromptText="Select or drag and drop integration bundles"
          onChange={handleFileChange}
          isInvalid={checkCompleted && isInvalid}
          onBlur={() => setBlurred(true)}
        />
      </EuiFormRow>
    </EuiForm>
  );
};

export const IntegrationUploadFlyout = ({ onClose }: { onClose: () => void }) => {
  const [bundle, setBundle] = useState(null as File | null);
  const { setToast } = useToast();
  const [isInvalid, setIsInvalid] = useState(true);

  return (
    <EuiFlyout onClose={onClose} size="s">
      <EuiFlyoutHeader>Upload Integrations</EuiFlyoutHeader>
      <EuiFlyoutBody>
        <IntegrationUploadPicker
          onFileSelected={setBundle}
          isInvalid={isInvalid}
          setIsInvalid={setIsInvalid}
        />
      </EuiFlyoutBody>
      <EuiFlyoutFooter>
        <EuiFlexGroup justifyContent="spaceBetween">
          <EuiFlexItem grow={false}>
            <EuiButton onClick={onClose}>Cancel</EuiButton>
          </EuiFlexItem>
          <EuiFlexItem grow={false}>
            <EuiButton
              fill
              disabled={isInvalid}
              onClick={async () => {
                const err = await uploadBundle(bundle);
                if (err == null) {
                  setToast('Successfully uploaded bundle', 'success');
                  onClose();
                } else {
                  setToast('Error uploading bundle', 'danger', err.message ?? '');
                }
              }}
            >
              Upload
            </EuiButton>
          </EuiFlexItem>
        </EuiFlexGroup>
      </EuiFlyoutFooter>
    </EuiFlyout>
  );
};
