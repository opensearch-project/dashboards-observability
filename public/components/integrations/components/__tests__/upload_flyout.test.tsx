/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { configure, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import React from 'react';
import { waitFor } from '@testing-library/react';
import { IntegrationUploadFlyout, IntegrationUploadPicker } from '../upload_flyout';

describe('Integration Upload Flyout', () => {
  configure({ adapter: new Adapter() });

  it('Renders integration upload flyout as expected', async () => {
    const wrapper = shallow(<IntegrationUploadFlyout onClose={() => {}} />);

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });

  it('Renders the clean integration picker as expected', async () => {
    const wrapper = shallow(
      <IntegrationUploadPicker
        isInvalid={true}
        setIsInvalid={(_) => {}}
        onFileSelected={(_) => {}}
      />
    );

    await waitFor(() => {
      expect(wrapper).toMatchSnapshot();
    });
  });
});
