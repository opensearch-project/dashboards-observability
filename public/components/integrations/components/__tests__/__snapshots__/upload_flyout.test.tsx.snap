// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Upload Flyout Renders integration upload flyout as expected 1`] = `
<EuiFlyout
  onClose={[Function]}
  size="s"
>
  <EuiFlyoutHeader>
    <EuiText
      size="s"
    >
      <h2>
        Upload Integrations
      </h2>
    </EuiText>
  </EuiFlyoutHeader>
  <EuiFlyoutBody>
    <IntegrationUploadPicker
      isInvalid={true}
      onFileSelected={[Function]}
      setIsInvalid={[Function]}
    />
  </EuiFlyoutBody>
  <EuiFlyoutFooter>
    <EuiFlexGroup
      justifyContent="spaceBetween"
    >
      <EuiFlexItem
        grow={false}
      >
        <EuiSmallButton
          onClick={[Function]}
        >
          Cancel
        </EuiSmallButton>
      </EuiFlexItem>
      <EuiFlexItem
        grow={false}
      >
        <EuiSmallButton
          disabled={true}
          fill={true}
          onClick={[Function]}
        >
          Upload
        </EuiSmallButton>
      </EuiFlexItem>
    </EuiFlexGroup>
  </EuiFlyoutFooter>
</EuiFlyout>
`;

exports[`Integration Upload Flyout Renders the clean integration picker as expected 1`] = `
<EuiForm>
  <EuiCompressedFormRow
    describedByIds={Array []}
    display="rowCompressed"
    error={
      Array [
        "Must be an ndjson bundle of integration templates",
      ]
    }
    fullWidth={false}
    hasChildLabel={true}
    hasEmptyLabelSpace={false}
    isInvalid={false}
    label="Select file"
    labelType="label"
  >
    <EuiCompressedFilePicker
      compressed={true}
      display="large"
      id="integrationBundlePicker"
      initialPromptText="Select or drag and drop integration bundles"
      isInvalid={false}
      onBlur={[Function]}
      onChange={[Function]}
    />
  </EuiCompressedFormRow>
</EuiForm>
`;
